break execute_model in gpu_model_runner into sub-functions for custom scopes

bangshengtang opened the pull request:
Summary:
break down the main execute_model() function into multiple parts so that we can have better latency breakdown in profiles:
- preprocess
- forward
- postprocess
- bookkeep (includes sync)
- draft (if spec decoding is enabled)

this is meant to be a refactor + some new function scopes, there's no functionality difference

Test Plan:
CI

run the vLLM predictor runbook and take a trace
<img width="2566" height="666" alt="image" src="https://github.com/user-attachments/assets/115f06cc-1c78-458d-ad84-bf9c185920ac" />

Differential Revision: D81009244

houseroad commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -14,6 +14,7 @@
 import torch
 import torch.distributed
 import torch.nn as nn
+from torch.autograd.profiler import record_function
```
let's use something like this?

```
def record_function_or_nullcontext(name: str, profiling: bool) -> ContextManager:
    if profiling:
        return record_function(name)
    else:
        return contextlib.nullcontext()
```

We can use some engine arg or ENV var to control it.

houseroad reviewed (approved):
I like this move. It will make the profiling result easier to parse. As by product, it also break long function into smaller pieces.

houseroad commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1726,20 +1667,153 @@ def execute_model(
+            )
+
+        with record_function_or_nullcontext("Postprocess"):
+            if self.use_aux_hidden_state_outputs:
```
later we can move this into something like _post_processing.

houseroad commented on vllm/v1/utils.py:
```diff
@@ -355,3 +359,10 @@ def report_usage_stats(
         })
+
+
+def record_function_or_nullcontext(name: str) -> AbstractContextManager:
+    if envs.VLLM_CUSTOM_SCOPES_FOR_PROFILING:
+        return record_function(name)
+    else:
+        return contextlib.nullcontext()
```
For use case for extreme perf, we would like to avoid additional overhead. Although, this should be light.

bangshengtang committed:
break execute_model in gpu_model_runner into sub-functions for custom scopes (#24265)

Summary:

break down the main execute_model() function into multiple parts so that we can have better latency breakdown in profiles:
- preprocess
- forward
- postprocess
- bookkeep (includes sync)
- draft (if spec decoding is enabled)

this is meant to be a refactor + some new function scopes, there's no functionality difference

Test Plan:
CI

run the vLLM predictor runbook and take a trace
{F1981500506}

Reviewed By: houseroad, frank-wei

Differential Revision: D81009244

bangshengtang committed:
Merge branch 'main' into export-D81009244

bangshengtang committed:
Merge branch 'main' into export-D81009244
